lib/Mojo/Message.pm

```diff
@@ -132,7 +132,8 @@ sub is_limit_exceeded { !!shift->{limit} }
 sub json {
   my ($self, $pointer) = @_;
   return undef if $self->content->is_multipart;
-  my $data = $self->{json} //= j($self->body);
+  my $charset = $self->body_params->charset || $self->default_charset;
```

Comment:
I wonder what the performance cost of this is.

Reply:
You are right, `$self->content->charset` is the way to go.